### PR TITLE
Fix NullPointerException on tomcat

### PR DESCRIPTION
**Pull request by laurit**
Resolves https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/3673
We store response in a request attribute and use it later to add async completion listener. Currently on tomcat we touch request a bit too early when it ins't fully initialized. Either `org.apache.coyote.Request` doesn't have a reference to `HttpServletRequest` in which case we just don't link response to request or `org.apache.coyote.Request` has reference to `HttpServletRequest` which is empty (request was recycled, reusing a previous request) in which case depending on tomcat version we could get a `NullPointerException`.

### CODE REVIEW

File: instrumentation/servlet/servlet-3.0/library/src/main/java/io/opentelemetry/instrumentation/servlet/v3_0/Servlet3Accessor.java [hunk 1]
```diff
@@ -31,8 +31,6 @@ public void addRequestAsyncListener(
       request
           .getAsyncContext()
           .addListener(new Listener(listener), request, (HttpServletResponse) response);
-    } else {
```
Comment (laurit) [inline instrumentation/servlet/servlet-3.0/library/src/main/java/io/opentelemetry/instrumentation/servlet/v3_0/Servlet3Accessor.java:34]: removed this branch to make tests fail when `response` is null

File: instrumentation/tomcat/tomcat-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat/common/TomcatServerHandlerInstrumentation.java [hunk 2]
```diff
@@ -36,6 +40,14 @@ public void transform(TypeTransformer transformer) {
             .and(named("service"))
             .and(takesArgument(0, named("org.apache.coyote.Request")))
             .and(takesArgument(1, named("org.apache.coyote.Response"))),
-        adviceClassName);
+        handlerAdviceClassName);
+
+    transformer.applyAdviceToMethod(
+        isMethod()
+            .and(named("postParseRequest"))
```
Comment (mateuszrzeszutek) [inline instrumentation/tomcat/tomcat-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat/common/TomcatServerHandlerInstrumentation.java:47]: `Adapter` doesn't have this method. Is there any reason why we're not instrumenting `CoyoteAdapter`?

**Review by mateuszrzeszutek — APPROVED** [review]

File: instrumentation/tomcat/tomcat-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat/common/TomcatServerHandlerInstrumentation.java (hunk 2 above)
Comment (laurit) [reply instrumentation/tomcat/tomcat-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat/common/TomcatServerHandlerInstrumentation.java:46]: Thanks. Didn't even realize that it wasn't using `CoyoteAdapter`.

**Review by trask — APPROVED** [review]
thx

**Pull request merged**